### PR TITLE
[BE1] validate sum of transaction outputs

### PR DESCRIPTION
**Pull request by lourkeur**
Maxime requested I take care of this. Corresponds to #1077.

There's a transparent type alias for Uint53 along with a summation operation.

We check sum of outputs because it's possible in a stateless way. Conveniently, it's the one you should check to avoid ex-nihilo money bugs.

### CODE REVIEW

File: be1-go/message/test/messagedata/coin_post_transaction_test.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ func Test_Coin_Post_Transaction(t *testing.T) {
 	require.Equal(t, "P2PKH", msg.Transaction.Inputs[0].Script.Type)
 	require.Equal(t, "oKHk3AivbpNXk_SfFcHDaVHcCcY8IBfHE7auXJ7h4ms=", msg.Transaction.Inputs[0].Script.PubKey)
 	require.Equal(t, "6hY16bDKnb7bRA5j7IMDR1CJDqAJKOLuQRgKxdpYQIrtSTVTRjo5jigqPhYQEPcK5a1WAF86V739ENFnlp6YCw==", msg.Transaction.Inputs[0].Script.Sig)
-	require.Equal(t, 32, msg.Transaction.Outputs[0].Value)
+	require.EqualValues(t, 32, msg.Transaction.Outputs[0].Value)
```
Comment (Tesa27) [inline be1-go/message/test/messagedata/coin_post_transaction_test.go:39]: Why is this necessary, is there a limitation of Equal function?

**Review by Tesa27 — COMMENTED** [review]
Looks good, just a small question.

File: be1-go/message/test/messagedata/coin_post_transaction_test.go (hunk 1 above)
Comment (lourkeur) [reply be1-go/message/test/messagedata/coin_post_transaction_test.go:39]: yes, Equal treats `uint64(32)` and `int(32)` as different. it's either that or changing the second argument

Comment (Tesa27) [reply be1-go/message/test/messagedata/coin_post_transaction_test.go:39]: Okay okay, agree then

File: be1-go/message/messagedata/coin_post_transaction.go [hunk 2]
```diff
@@ -52,17 +53,27 @@ type UnlockScript struct {
...
 // Verify verifies that the PostTransaction message is valid
 func (message PostTransaction) Verify() error {
-	for _, out := range message.Transaction.Outputs {
-		if out.Value < 0 {
-			return answer.NewErrorf(-4, "transaction output value is %d, "+
-				"shouldn't be negative", out.Value)
-		}
+	_, err := message.Transaction.SumOutputs()
+	if err != nil {
+		return err
```
Comment (Robingoumaz) [inline be1-go/message/messagedata/coin_post_transaction.go:72]: You should wrap the error

**Review by Robingoumaz — CHANGES_REQUESTED** [review]
Nice, some minor changes needed

**Comment by lourkeur** [comment]
Will close #988

**Review by Robingoumaz — DISMISSED** [review]

File: be1-go/channel/coin/uint53/mod.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
+  Operations on 53-bits non-negative integers.
+
+  These values can be handled without loss of precision in IEEE754 64-bit
+  floating numbers, and thus in JSON.
+*/
```
Comment (nkcr) [inline be1-go/channel/coin/uint53/mod.go:6]: You must use `//` and the comment should start with "Package uint53 ...".
See https://go.dev/doc/comment. Also be sure comment lines are wrapped in 80 chars.

File: be1-go/channel/coin/uint53/mod.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+package uint53
+
+import (
+	"errors"
+)
+
+// Internal representation in this implementation
```
Comment (nkcr) [inline be1-go/channel/coin/uint53/mod.go:13]: Comments on public types and functions must "...start with complete sentences naming the declared symbol".
We always start the comment with the symbol's name:

```go
// Uint53 defines the internal representation in this implementation
```

Same bellow.

File: be1-go/channel/coin/uint53/mod.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+const (
+	MaxUint53 Uint53 = 0x1F_FF_FF_FF_FF_FF_FF
+)
+
+// Check whether the value is in range
+func InRange(a Uint53) bool {
+	return a <= MaxUint53
+}
+
+// Compute the sum of two Uint53 values, or indicate an overflow if the sum is too large.
```
Comment (nkcr) [inline be1-go/channel/coin/uint53/mod.go:26]:
```suggestion
// SafePlus compute the sum of two Uint53 values, or indicate an overflow if the
// sum is too large.
```

File: be1-go/message/messagedata/coin_post_transaction.go [hunk 6]
```diff
@@ -52,17 +53,28 @@ type UnlockScript struct {
 	Sig    string `json:"sig"`
 }
 
+// SumOutputs computes the sum of the amounts in all outputs.
+// It may signal an overflow error
+func (transaction Transaction) SumOutputs() (acc uint53.Uint53, err error) {
+	for _, out := range transaction.Outputs {
+		acc, err = uint53.SafePlus(acc, out.Value)
+		if err != nil {
+			return 0, err
```
Comment (nkcr) [inline be1-go/message/messagedata/coin_post_transaction.go:62]: we always wrap errors:

```go
xerrors.Errorf("failed to perform uint53 addition: %v", err)
```

File: be1-go/message/messagedata/coin_post_transaction.go [hunk 7]
```diff
@@ -52,17 +53,28 @@ type UnlockScript struct {
 	Sig    string `json:"sig"`
 }
 
+// SumOutputs computes the sum of the amounts in all outputs.
+// It may signal an overflow error
+func (transaction Transaction) SumOutputs() (acc uint53.Uint53, err error) {
```
Comment (nkcr) [inline be1-go/message/messagedata/coin_post_transaction.go:58]: we tend to never use named return values:

```suggestion
func (transaction Transaction) SumOutputs() (uint53.Uint53, error) {
  var acc uint53.Uint53
  var err error
```

File: be1-go/channel/coin/uint53/mod.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+func SafePlus(a, b Uint53) (r Uint53, err error) {
+	if !InRange(a) || !InRange(b) {
+		panic("Uint53.SafePlus: argument out of range")
```
Comment (nkcr) [inline be1-go/channel/coin/uint53/mod.go:29]: I would not panic and return an error. We probably don't want to server to crash for this reason.

File: be1-go/channel/coin/uint53/mod_test.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	require.Equal(t, Uint53(1<<53-1), MaxUint53)
+}
+
+func Test_InRange_Examples(t *testing.T) {
+	require.True(t, InRange(0))
+	require.True(t, InRange(MaxUint53))
+	require.False(t, InRange(MaxUint53+1))
+}
+
+func Test_SafePlus_Examples(t *testing.T) {
```
Comment (nkcr) [inline be1-go/channel/coin/uint53/mod_test.go:19]: you should also add cases where an error is expected: 

```go
SafePlus(MaxUint53, 1)
SafePlus(MaxUint53+1, 0)
```

Comment (lourkeur) [reply be1-go/channel/coin/uint53/mod_test.go:19]: `SafePlus(MaxUint53, 1)` is already a test case, `SafePlus(MaxUint53+1, 0)` violates pre-conditions.

File: be1-go/channel/coin/uint53/mod.go (hunk 8 above)
Comment (lourkeur) [reply be1-go/channel/coin/uint53/mod.go:29]: I look at this more as a violation of pre-conditions than something that goes wrong with the operation itself, that said it's the only panic in the codebase so I will change it.

File: be1-go/channel/coin/uint53/mod_test.go (hunk 9 above)
Comment (lourkeur) [reply be1-go/channel/coin/uint53/mod_test.go:19]: added at the bottom

**Review by nkcr — APPROVED** [review]
Great, thanks 👍
